controller/utils.go

```diff
@@ -61,165 +60,182 @@ func SegmentAndReplacePods(
 		"scheduler": schedulerName,
 	})
 
+	ctx, cancel := context.WithDeadline(context.Background(), willTimeoutAt)
+	defer cancel()
+
+	inRollingUpdate := operationManager != nil
+
+	if inRollingUpdate {
+		go watchOperation(ctx, cancel, operationManager)
+	}
+
 	// segment pods in chunks
 	podChunks := segmentPods(pods, maxSurge)
 
 	for i, chunk := range podChunks {
 		l.Debugf("updating chunk %d: %v", i, names(chunk))
 
 		// replace chunk
-		timedout, canceled, errored := replacePodsAndWait(
-			l,
+		err := replacePodsAndWait(
+			ctx,
+			logger,
 			roomManager,
 			mr,
 			clientset,
 			db,
 			redisClient,
-			willTimeoutAt,
 			configYAML,
 			chunk,
 			scheduler,
-			operationManager,
-			clock,
+			inRollingUpdate,
 			goroutinePoolSize,
 		)
 
-		if timedout {
-			timeoutErr = errors.New("timedout waiting rooms to be replaced, rolled back")
-			l.WithError(timeoutErr).Error("error replacing chunk of pods")
-			break
+		if err != nil {
+			l.WithError(err).Error("error replacing chunk of pods")
+			return nil, nil, err
 		}
+	}
 
-		if canceled {
-			cancelErr = errors.New("operation was canceled, rolled back")
-			l.WithError(cancelErr).Error("error replacing chunk of pods")
-			break
-		}
+	cancelReason := ctx.Err()
 
-		if errored != nil {
-			err = errored
-			l.WithError(errored).Error("error replacing chunk of pods")
-			break
-		}
+	if cancelReason == context.DeadlineExceeded {
+		timeoutErr = errors.New("timedout waiting rooms to be replaced, rolled back")
+		l.WithError(timeoutErr).Error("operation timed out while replacing chunk of pods")
+	} else if cancelReason == context.Canceled {
+		cancelErr = errors.New("operation was canceled, rolled back")
+		l.WithError(cancelErr).Error("operation canceled while error replacing chunk of pods")
 	}
 
 	return timeoutErr, cancelErr, err
 }
 
+func watchOperation(ctx context.Context, cancel func(), operationManager *models.OperationManager) {
+	canceled, err := operationManager.WasCanceled()
+	if canceled && err  == nil {
+		cancel()
+		return
+	}
+
+	ticker := time.NewTicker(10 * time.Second)
```

Comment:
Isn't 10 seconds too long? For the user the cancel operation will have a 10 second delay._

Reply:
Maybe putting this value in a config also helps.

Reply:
I just used the same value that was used before on the loop inside OperationManager but I can read this from a config.